### PR TITLE
[PowerPC] Fix vector_shuffle combines when inputs are scalar_to_vector of differing types.

### PR DESCRIPTION
**Pull request by amy-kwan**
This patch fixes the combines for vector_shuffles when either or both of its left
and right hand side inputs are scalar_to_vector nodes.

Previously, when both left and right side inputs are scalar_to_vector nodes, the
current combine could not handle this situation, as the shuffle mask was updated
incorrectly. To temporarily solve this solution, this combine was simply disabled
and not performed.

Now, not only does this patch aim to resolve the previous issue of the incorrect
shuffle mask adjustments respectively, but it also updates any test cases that are
affected by this change.

Patch migrated from https://reviews.llvm.org/D130487.

### CODE REVIEW

**Comment by amy-kwan** [comment]
Ping.

**Comment by amy-kwan** [comment]
Ping.

**Comment by amy-kwan** [comment]
Ping.

**Comment by amy-kwan** [comment]
Ping on this review once again :)

File: llvm/lib/Target/PowerPC/PPCISelLowering.cpp [hunk 1]
```diff
@@ -15238,16 +15238,18 @@ static SDValue isScalarToVec(SDValue Op) {
...
-static void fixupShuffleMaskForPermutedSToV(SmallVectorImpl<int> &ShuffV,
-                                            int LHSMaxIdx, int RHSMinIdx,
-                                            int RHSMaxIdx, int HalfVec,
-                                            unsigned ValidLaneWidth,
-                                            const PPCSubtarget &Subtarget) {
+static void fixupShuffleMaskForPermutedSToV(
+    SmallVectorImpl<int> &ShuffV, int LHSFirstElt, int LHSLastElt,
+    int RHSFirstElt, int RHSLastElt, int HalfVec, unsigned LHSNumValidElts,
+    unsigned RHSNumValidElts, const PPCSubtarget &Subtarget) {
   for (int i = 0, e = ShuffV.size(); i < e; i++) {
```
Comment (redstar) [inline llvm/lib/Target/PowerPC/PPCISelLowering.cpp:15245]: Sorry,  a nit: according to the coding standard, [variable names should be uppercase](https://llvm.org/docs/CodingStandards.html#name-types-functions-variables-and-enumerators-properly). In addition, [preincrement is preferred](https://llvm.org/docs/CodingStandards.html#prefer-preincrement):
 ```suggestion
  for (int I = 0, E = ShuffV.size(); I < E; ++I) {
```

File: llvm/lib/Target/PowerPC/PPCISelLowering.cpp [hunk 2]
```diff
@@ -15286,6 +15288,25 @@ static SDValue getSToVPermuted(SDValue OrigSToV, SelectionDAG &DAG,
                      OrigSToV.getOperand(0));
 }
 
+static bool isShuffleMaskInRange(const SmallVectorImpl<int> &ShuffV,
+                                 int HalfVec, int LHSLastElementDefined,
+                                 int RHSLastElementDefined) {
+  for (int i : seq<int>(0, ShuffV.size())) {
```
Comment (redstar) [inline llvm/lib/Target/PowerPC/PPCISelLowering.cpp:15294]: Same as above:
```suggestion
  for (int I : seq<int>(0, ShuffV.size())) {
```

File: llvm/lib/Target/PowerPC/PPCISelLowering.cpp [hunk 3]
```diff
@@ -15371,33 +15381,52 @@ SDValue PPCTargetLowering::combineVectorShuffle(ShuffleVectorSDNode *SVN,
...
         return Res;
       // Set up the values for the shuffle vector fixup.
-      LHSMaxIdx = NumEltsOut / NumEltsIn;
+      LHSNumValidElts =
+          LHSScalarSize / LHS.getValueType().getScalarSizeInBits();
+      // The last element that comes from the LHS. For example:
+      // (shuff (s_to_v i32), (bitcast (s_to_v i64), v4i32), ...)
+      // The last element that comes from the LHS is actually 0, not 3
+      // because elements 1 and higher of a scalar_to_vector are undefined.
+      LHSLastElt = LHSScalarSize / (ShuffleEltWidth + 1);
```
Comment (diggerlin) [inline llvm/lib/Target/PowerPC/PPCISelLowering.cpp:15394]: not sure why `ShuffleEltWidth + 1` in `LHSLastElt = LHSScalarSize / (ShuffleEltWidth + 1); ` , can you explain it ?

File: llvm/lib/Target/PowerPC/PPCISelLowering.cpp [hunk 4]
```diff
@@ -15371,33 +15381,52 @@ SDValue PPCTargetLowering::combineVectorShuffle(ShuffleVectorSDNode *SVN,
...
-      RHSMaxIdx = NumEltsOut / NumEltsIn + RHSMinIdx;
+      RHSNumValidElts =
+          RHSScalarSize / RHS.getValueType().getScalarSizeInBits();
+      // The last element that comes from the RHS. For example:
+      // (shuff (s_to_v i32), (bitcast (s_to_v i64), v4i32), ...)
+      // The last element that comes from the RHS is actually 5, not 7
+      // because elements 1 and higher of a scalar_to_vector are undefined.
+      // It is also not 4 because the original scalar_to_vector is wider and
+      // actually contains two i32 elements.
+      RHSLastElt = RHSScalarSize / (ShuffleEltWidth + 1) + RHSFirstElt;
```
Comment (diggerlin) [inline llvm/lib/Target/PowerPC/PPCISelLowering.cpp:15412]: most of code 
```
if (SToVLHS) {
  ... 
}
```

as same with 
```
if (SToVRHS) {
... 
}
```

can you add a help function or lambda function to simply the code ?

File: llvm/lib/Target/PowerPC/PPCISelLowering.cpp (hunk 3 above)
Comment (amy-kwan) [reply llvm/lib/Target/PowerPC/PPCISelLowering.cpp]: This is just for adjustment for the computation to ensure we get the correct last element.

For example,
```
          t36: i32,ch = load<(load (s32) from %ir.a)> t0, t2, undef:i64
        t37: v4i32 = scalar_to_vector t36
      t38: v16i8 = bitcast t37
            t4: i64,ch = CopyFromReg t0, Register:i64 %1
          t39: i64,ch = load<(load (s64) from %ir.b)> t0, t4, undef:i64
        t40: v2i64 = scalar_to_vector t39
      t41: v16i8 = bitcast t40
    t17: v16i8 = vector_shuffle<0,1,16,17,18,19,20,21,22,23,u,u,u,u,u,u> t38, t41
```
For this, I believe we should have:
```
LHSFirstElt = 0
LHSLastElt = 3
RHSFirstElt = 16
RHSLastElt = 23
```
However, if we did `LHSLastElt = LHSScalarSize / (ShuffleEltWidth);`, we would have ` 32 / 8 = 4`, which isn't correct. If we did `32 / 8 + 1 = 32 / 9`, we would get 3, which is what we expect.

File: llvm/lib/Target/PowerPC/PPCISelLowering.cpp [hunk 5]
```diff
@@ -15519,33 +15556,33 @@ SDValue PPCTargetLowering::combineVectorShuffle(ShuffleVectorSDNode *SVN,
...
     // The minimum and maximum indices that correspond to element zero for both
     // the LHS and RHS are computed and will control which shuffle mask entries
     // are to be changed. For example, if the RHS is permuted, any shuffle mask
-    // entries in the range [RHSMinIdx,RHSMaxIdx) will be adjusted.
-    fixupShuffleMaskForPermutedSToV(ShuffV, LHSMaxIdx, RHSMinIdx, RHSMaxIdx,
-                                    HalfVec, ValidLaneWidth, Subtarget);
+    // entries in the range [RHSFirstElt,RHSLastElt] will be adjusted.
+    fixupShuffleMaskForPermutedSToV(
+        ShuffV, LHSFirstElt, LHSLastElt, RHSFirstElt, RHSLastElt, HalfVec,
+        LHSNumValidElts, RHSNumValidElts, Subtarget);
```
Comment (maryammo) [inline llvm/lib/Target/PowerPC/PPCISelLowering.cpp:15585]: it seems LHSNumValidElts and RHSNumValidElts are both equal to HalfVec, if so why passing 2 separate parameter for them?

File: llvm/lib/Target/PowerPC/PPCISelLowering.cpp [hunk 6]
```diff
@@ -15519,33 +15556,33 @@ SDValue PPCTargetLowering::combineVectorShuffle(ShuffleVectorSDNode *SVN,
...
     // Fix up the shuffle mask to reflect where the desired element actually is.
     // The minimum and maximum indices that correspond to element zero for both
     // the LHS and RHS are computed and will control which shuffle mask entries
     // are to be changed. For example, if the RHS is permuted, any shuffle mask
-    // entries in the range [RHSMinIdx,RHSMaxIdx) will be adjusted.
-    fixupShuffleMaskForPermutedSToV(ShuffV, LHSMaxIdx, RHSMinIdx, RHSMaxIdx,
-                                    HalfVec, ValidLaneWidth, Subtarget);
+    // entries in the range [RHSFirstElt,RHSLastElt] will be adjusted.
+    fixupShuffleMaskForPermutedSToV(
+        ShuffV, LHSFirstElt, LHSLastElt, RHSFirstElt, RHSLastElt, HalfVec,
```
Comment (maryammo) [inline llvm/lib/Target/PowerPC/PPCISelLowering.cpp:15584]: same question about LHSFirstElt?

File: llvm/lib/Target/PowerPC/PPCISelLowering.cpp [hunk 7]
```diff
@@ -15519,33 +15556,33 @@ SDValue PPCTargetLowering::combineVectorShuffle(ShuffleVectorSDNode *SVN,
...
+      if (!IsLittleEndian && LHSScalarSize >= 64)
         return Res;
-      // Set up the values for the shuffle vector fixup.
-      LHSMaxIdx = NumEltsOut / NumEltsIn;
-      SToVLHS = getSToVPermuted(SToVLHS, DAG, Subtarget);
-      if (SToVLHS.getValueType() != LHS.getValueType())
-        SToVLHS = DAG.getBitcast(LHS.getValueType(), SToVLHS);
-      LHS = SToVLHS;
+      LHS = generateSToVPermutedForVecShuffle(
+          LHSScalarSize, ShuffleEltWidth, LHSNumValidElts, LHSFirstElt,
```
[the remark this answers is not in the document]
Comment (amy-kwan) [reply llvm/lib/Target/PowerPC/PPCISelLowering.cpp:15563]: Yeah, the first element of LHS will always be zero. Originally, I didn't have the `generateSToVPermutedForVecShuffle()` function, but I added it due to the comment I received in  https://github.com/llvm/llvm-project/pull/80784/commits/606ea91671401b0fbdd4a2e183587852a777a0b9#r1560041740 in attempts to common up some code.

The first element of the LHS actually doesn't matter when I'm generating the `SCALAR_TO_VECTOR_PERMUTED` within `generateSToVPermutedForVecShuffle()`, so I could just pass in zero, but where it matters is at the RHS.

I chose to still pass in `LHSFirstElt` here just to be explicit in the code that we're dealing with the first element of the LHS.

File: llvm/lib/Target/PowerPC/PPCISelLowering.cpp (hunk 6 above)
Comment (amy-kwan) [reply llvm/lib/Target/PowerPC/PPCISelLowering.cpp:15584]: Although the LHSFirstElt and RHSFirstElt values do not change, I think it would be good to be explicit to mention them here so it's obvious to the reader of the code what values are needed when we need to fix up the shuffle mask (after generating the `SCALAR_TO_VECTOR_PERMUTED` node).

File: llvm/lib/Target/PowerPC/PPCISelLowering.cpp (hunk 5 above)
Comment (amy-kwan) [reply llvm/lib/Target/PowerPC/PPCISelLowering.cpp:15585]: That is correct that they are both equal to `HalfVec` initially, but they can change within `generateSToVPermutedForVecShuffle()` since one of the parameters to this function is `unsigned &NumValidElts` so I require them to be separate variables.

**Comment by amy-kwan** [comment]
Ping. :-)

File: llvm/test/CodeGen/PowerPC/v16i8_scalar_to_vector_shuffle.ll [hunk 8]
```diff
@@ -266,56 +266,54 @@ entry:
...
 ;
 ; CHECK-BE-P8-LABEL: test_v16i8_v8i16:
 ; CHECK-BE-P8:       # %bb.0: # %entry
-; CHECK-BE-P8-NEXT:    sldi r4, r4, 56
-; CHECK-BE-P8-NEXT:    sldi r3, r3, 48
-; CHECK-BE-P8-NEXT:    mtvsrd v2, r4
-; CHECK-BE-P8-NEXT:    mtvsrd v3, r3
-; CHECK-BE-P8-NEXT:    vmrghh v2, v2, v3
+; CHECK-BE-P8-NEXT:    mtvsrwz v3, r3
+; CHECK-BE-P8-NEXT:    addis r3, r2, .LCPI3_0@toc@ha
```
Comment (diggerlin) [inline llvm/test/CodeGen/PowerPC/v16i8_scalar_to_vector_shuffle.ll:284]: I am curious . there is no read data from TC entry, but the right has the read data from TC entry, but the PR look not related to TC entry ?

File: llvm/test/CodeGen/PowerPC/p8-scalar_vector_conversions.ll [hunk 9]
```diff
@@ -2505,11 +2505,9 @@ define <2 x i64> @buildi2(i64 %arg, i32 %arg1) {
 ;
 ; CHECK-LE-LABEL: buildi2:
 ; CHECK-LE:       # %bb.0: # %entry
-; CHECK-LE-NEXT:    mtfprd f0, r4
+; CHECK-LE-NEXT:    mtfprwz f0, r4
```
[the remark this answers is not in the document]
Comment (amy-kwan) [reply llvm/test/CodeGen/PowerPC/p8-scalar_vector_conversions.ll:2508]: Unless I am misunderstanding, the `mtfprwz f0, r4` represents the second `i32 %arg1` in this function, so I think this should be fine.

**Comment by amy-kwan** [comment]
Ping.

File: llvm/test/CodeGen/PowerPC/v2i64_scalar_to_vector_shuffle.ll [hunk 10]
```diff
@@ -102,42 +102,42 @@ define <2 x i64> @test_none_v16i8(i8 %arg1, ptr nocapture noundef readonly %b) {
 ; CHECK-LE-P8-NEXT:    lxvd2x vs0, 0, r4
 ; CHECK-LE-P8-NEXT:    xxswapd v2, vs0
 ; CHECK-LE-P8-NEXT:    mtfprd f0, r3
-; CHECK-LE-P8-NEXT:    xxmrgld v2, vs0, v2
+; CHECK-LE-P8-NEXT:    xxpermdi v2, vs0, v2, 1
```
Comment (diggerlin) [inline llvm/test/CodeGen/PowerPC/v2i64_scalar_to_vector_shuffle.ll:105]: according to the IR.

```
 159 entry:
 160   %lhs.tmp = insertelement <16 x i8> undef, i8 %arg1, i32 0
 161   %lhs = bitcast <16 x i8> %lhs.tmp to <2 x i64>
 162   %rhs = load <2 x i64>, ptr %b, align 4
 163   %shuffle = shufflevector <2 x i64> %rhs, <2 x i64> %lhs, <2 x i32> <i32 0, i32 2>
 164   ret <2 x i64> %shuffle
 165 }
```
line 163 indicated that the  %shuffle consist from  first element of the first vector (%rhs[0]) and  the first element of the second vector (%lhs[0]).

Mask index 0 refers to the first element of the first vector (%rhs[0]).
Mask index 2 refers to the first element of the second vector (%lhs[0]).

it should be `CHECK-LE-P8-NEXT: xxpermdi v2, vs0, v2, 0`  or `xxmrgld v2, vs0, v2` , Do I understand error ?

File: llvm/test/CodeGen/PowerPC/v2i64_scalar_to_vector_shuffle.ll [hunk 11]
```diff
@@ -170,42 +170,42 @@ define <2 x i64> @test_v16i8_none(i8 %arg1, ptr nocapture noundef readonly %b) {
 ; CHECK-LE-P8-NEXT:    lxvd2x vs0, 0, r4
 ; CHECK-LE-P8-NEXT:    xxswapd v2, vs0
 ; CHECK-LE-P8-NEXT:    mtfprd f0, r3
-; CHECK-LE-P8-NEXT:    xxmrgld v2, v2, vs0
+; CHECK-LE-P8-NEXT:    xxpermdi v2, v2, vs0, 2
```
Comment (diggerlin) [inline llvm/test/CodeGen/PowerPC/v2i64_scalar_to_vector_shuffle.ll:173]: it should be `xxpermdi v2, v2, vs0, 0` ?

File: llvm/test/CodeGen/PowerPC/v16i8_scalar_to_vector_shuffle.ll [hunk 12]
```diff
@@ -1669,53 +1656,54 @@ entry:
...
 ; CHECK-AIX-64-P9:       # %bb.0: # %entry
-; CHECK-AIX-64-P9-NEXT:    mtvsrws v2, r3
-; CHECK-AIX-64-P9-NEXT:    sldi r3, r4, 48
-; CHECK-AIX-64-P9-NEXT:    mtvsrd v3, r3
-; CHECK-AIX-64-P9-NEXT:    vmrghh v2, v2, v3
+; CHECK-AIX-64-P9-NEXT:    mtfprwz f0, r3
+; CHECK-AIX-64-P9-NEXT:    ld r3, L..C10(r2) # %const.0
+; CHECK-AIX-64-P9-NEXT:    mtvsrwz v2, r4
+; CHECK-AIX-64-P9-NEXT:    lxv vs1, 0(r3)
+; CHECK-AIX-64-P9-NEXT:    xxperm v2, vs0, vs1
```
Comment (diggerlin) [inline llvm/test/CodeGen/PowerPC/v16i8_scalar_to_vector_shuffle.ll:1706]: I searched from the ISA, it looks xxperm has four operands ?
e.g. :
xxperm vT, vLD, vT, vPCV

File: llvm/lib/Target/PowerPC/PPCISelLowering.cpp [hunk 13]
```diff
@@ -15771,33 +15808,33 @@ SDValue PPCTargetLowering::combineVectorShuffle(ShuffleVectorSDNode *SVN,
...
+          LHSLastElt, LHS, SToVLHS, DAG, Subtarget);
     }
     if (SToVRHS) {
-      if (!IsLittleEndian && SToVRHS.getValueType().getScalarSizeInBits() >= 64)
+      int RHSScalarSize = SToVRHS.getValueType().getScalarSizeInBits();
+      if (!IsLittleEndian && RHSScalarSize >= 64)
```
Comment (diggerlin) [inline llvm/lib/Target/PowerPC/PPCISelLowering.cpp:15820]: I think if we hit run   `LHS = generateSToVPermutedForVecShuffle`
but  `!IsLittleEndian && RHSScalarSize >= 64` is true

it still need to run 
function `isShuffleMaskInRange`

File: llvm/lib/Target/PowerPC/PPCISelLowering.cpp [hunk 14]
```diff
@@ -15686,6 +15690,50 @@ static SDValue getSToVPermuted(SDValue OrigSToV, SelectionDAG &DAG,
                      OrigSToV.getOperand(0));
 }
 
+static bool isShuffleMaskInRange(const SmallVectorImpl<int> &ShuffV,
+                                 int HalfVec, int LHSLastElementDefined,
+                                 int RHSLastElementDefined) {
+  for (int I : seq<int>(0, ShuffV.size())) {
+    int Index = ShuffV[I];
```
Comment (redstar) [inline llvm/lib/Target/PowerPC/PPCISelLowering.cpp:15697]: `I` is only used to get the element out of `ShuffV`. That can be done easier:
```suggestion
  for (int Index : ShuffV) {
```

File: llvm/test/CodeGen/PowerPC/v16i8_scalar_to_vector_shuffle.ll (hunk 12 above)
Comment (amy-kwan) [reply llvm/test/CodeGen/PowerPC/v16i8_scalar_to_vector_shuffle.ll:1706]: In the ISA, the page for `xxperm` says:
```
VSX Vector Permute XX3-form
xxperm XT,XA,XB
```
All other LLVM LIT tests also have the same format of `xxperm` having three operands.

File: llvm/lib/Target/PowerPC/PPCISelLowering.cpp [hunk 15]
```diff
@@ -15686,6 +15690,50 @@ static SDValue getSToVPermuted(SDValue OrigSToV, SelectionDAG &DAG,
...
+  // For example:
+  // (shuff (s_to_v i32), (bitcast (s_to_v i64), v4i32), ...)
+  //
+  // For the LHS: The last element that comes from the LHS is actually 0, not 3
+  // because elements 1 and higher of a scalar_to_vector are undefined.
+  // For the RHS: The last element that comes from the RHS is actually 5, not 7
+  // because elements 1 and higher of a scalar_to_vector are undefined.
+  // It is also not 4 because the original scalar_to_vector is wider and
+  // actually contains two i32 elements.
+  LastElt = ScalarSize / (ShuffleEltWidth + 1) + FirstElt;
```
[the remark this answers is not in the document]
Comment (amy-kwan) [reply llvm/lib/Target/PowerPC/PPCISelLowering.cpp:15730]: This is explained in https://github.com/llvm/llvm-project/pull/80784#discussion_r1571032532.

File: llvm/lib/Target/PowerPC/PPCISelLowering.cpp (hunk 13 above)
Comment (amy-kwan) [reply llvm/lib/Target/PowerPC/PPCISelLowering.cpp:15820]: Do you mean, do we still need to run `isShuffleMaskInRange()` if we generate a `SCALAR_TO_VECTOR_PERMUTED` for the LHS, but not the RHS side?

I don't think so. I think this is due to the above comment:
```
    // On big endian systems, this only makes sense for element sizes smaller
    // than 64 bits since for 64-bit elements, all instructions already put
    // the value into element zero. Since scalar size of LHS and RHS may differ
    // after isScalarToVec, this should be checked using their own sizes.
```

File: llvm/lib/Target/PowerPC/PPCISelLowering.cpp (hunk 14 above)
Comment (amy-kwan) [reply llvm/lib/Target/PowerPC/PPCISelLowering.cpp:15697]: Great suggestion, thanks!

File: llvm/lib/Target/PowerPC/PPCISelLowering.cpp (hunk 15 above)
[the remark this answers is not in the document]
Comment (diggerlin) [reply llvm/lib/Target/PowerPC/PPCISelLowering.cpp:15730]: it get the correct value, but not easy to understand the purpose from the statement  `LastElt = ScalarSize / (ShuffleEltWidth + 1) + FirstElt`

I think the one will the same the value as above

`LastElt = ScalarSize < ShuffleEltWidth ?  ScalarSize / ShuffleEltWidth -1 + FirstElt :  FirstElt`

and it is easy to understand.

File: llvm/lib/Target/PowerPC/PPCISelLowering.cpp (hunk 13 above)
Comment (diggerlin) [reply llvm/lib/Target/PowerPC/PPCISelLowering.cpp:15820]: if (!IsLittleEndian && RHSScalarSize >= 64)
        return Res;
        
  the patch still generate LHS, which are not used actually.
  
  can we change the code to
  
  
     ```
 if (SToVLHS) {
      LHSScalarSize = SToVLHS.getValueType().getScalarSizeInBits();
      if (!IsLittleEndian && LHSScalarSize >= 64)
        return Res;
    }

    if (SToVRHS) {
      int RHSScalarSize = SToVRHS.getValueType().getScalarSizeInBits();
      if (!IsLittleEndian && RHSScalarSize >= 64)
        return Res;
    }

    if(LHSScalarSize != 0)
      LHS = generateSToVPermutedForVecShuffle(
          LHSScalarSize, ShuffleEltWidth, LHSNumValidElts, LHSFirstElt,
          LHSLastElt, LHS, SToVLHS, DAG, Subtarget);

   if (RHSScalarSize != 0 )
      RHS = generateSToVPermutedForVecShuffle(
          RHSScalarSize, ShuffleEltWidth, RHSNumValidElts, RHSFirstElt,
          RHSLastElt, RHS, SToVRHS, DAG, Subtarget);

    if (!isShuffleMaskInRange(ShuffV, HalfVec, LHSLastElt, RHSLastElt))
      return Res;
  
```

File: llvm/test/CodeGen/PowerPC/v2i64_scalar_to_vector_shuffle.ll (hunk 10 above)
Comment (amy-kwan) [reply llvm/test/CodeGen/PowerPC/v2i64_scalar_to_vector_shuffle.ll:105]: I think this should be resolved now from our offline discussion, but there was an issue in the previous codegen where we produced a `SCALAR_TO_VECTOR_PERMUTED` but we did not fix up the shuffle mask. This patch updates the shuffle mask, so this codegen should be correct.

File: llvm/test/CodeGen/PowerPC/v2i64_scalar_to_vector_shuffle.ll (hunk 11 above)
Comment (amy-kwan) [reply llvm/test/CodeGen/PowerPC/v2i64_scalar_to_vector_shuffle.ll:173]: I think this should be resolved now from our offline discussion, but there was an issue in the previous codegen where we produced a `SCALAR_TO_VECTOR_PERMUTED` but we did not fix up the shuffle mask. This patch updates the shuffle mask, so this codegen should be correct.

File: llvm/lib/Target/PowerPC/PPCISelLowering.cpp (hunk 15 above)
[the remark this answers is not in the document]
Comment (amy-kwan) [reply llvm/lib/Target/PowerPC/PPCISelLowering.cpp:15730]: Digger and I briefly discussed this and the suggestion should instead be:
```
LastElt = ScalarSize >  ShuffleEltWidth ? ScalarSize / ShuffleEltWidth -1 + FirstElt : FirstElt
```
I will test this and report back.

File: llvm/lib/Target/PowerPC/PPCISelLowering.cpp [hunk 16]
```diff
@@ -15687,16 +15687,20 @@ static SDValue isScalarToVec(SDValue Op) {
...
+    unsigned RHSNumValidElts, const PPCSubtarget &Subtarget) {
+  int LHSEltFixup =
+      Subtarget.isLittleEndian() ? HalfVec : HalfVec - LHSNumValidElts;
+  int RHSEltFixup =
+      Subtarget.isLittleEndian() ? HalfVec : HalfVec - RHSNumValidElts;
+  for (int I = 0, E = ShuffV.size(); I < E; ++I) {
+    int Idx = ShuffV[I];
+    if (Idx >= LHSFirstElt && Idx <= LHSLastElt)
+      ShuffV[I] += LHSEltFixup;
+    if (Idx >= RHSFirstElt && Idx <= RHSLastElt)
```
Comment (diggerlin) [inline llvm/lib/Target/PowerPC/PPCISelLowering.cpp:15702]: nit: change to 
  `else  if (Idx >= RHSFirstElt && Idx <= RHSLastElt)`

**Review by diggerlin — APPROVED** [review]
